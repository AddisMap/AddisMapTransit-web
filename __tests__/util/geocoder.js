import nock from 'nock'

import getGeocoder from '../../lib/util/geocoder'

function mockResponsePath (geocoder, file) {
  return `__tests__/test-utils/fixtures/geocoding/${geocoder}/${file}`
}

describe('geocoder', () => {
  const geocoders = [
    {
      apiKey: 'dummy-mapzen-key',
      baseUrl: 'https://ws-st.trimet.org/pelias/v1',
      type: 'PELIAS'
    }, {
      type: 'ARCGIS'
    }
  ]

  // nocks for ARCGIS
  const baseArcGisPath = '/arcgis/rest/services/World/GeocodeServer/'
  nock('https://geocode.arcgis.com')
    // autocomplete
    .get(`${baseArcGisPath}suggest`)
    .query(true)
    .replyWithFile(200, mockResponsePath('arcgis', 'suggest-response.json'))
    // reverse
    .get(`${baseArcGisPath}findAddressCandidates`)
    .query(true)
    .replyWithFile(200, mockResponsePath('arcgis', 'findAddressCandidates-response.json'))
    // search
    .get(`${baseArcGisPath}reverseGeocode`)
    .query(true)
    .replyWithFile(200, mockResponsePath('arcgis', 'reverseGeocode-response.json'))

  // nocks for PELIAS
  const basePeliasPath = '/pelias/v1/'
  nock('https://ws-st.trimet.org')
    // autocomplete
    .get(`${basePeliasPath}autocomplete`)
    .query(true)
    .replyWithFile(200, mockResponsePath('pelias', 'autocomplete-response.json'))
    // reverse
    .get(`${basePeliasPath}search`)
    .query(true)
    .replyWithFile(200, mockResponsePath('pelias', 'search-response.json'))
    // search
    .get(`${basePeliasPath}reverse`)
    .query(true)
    .replyWithFile(200, mockResponsePath('pelias', 'reverse-response.json'))

  geocoders.forEach(geocoder => {
    // the describe is in quotes to bypass a lint rule
    describe(`${geocoder.type}`, () => {
      it('should make autocomplete query', async () => {
        const result = await getGeocoder(geocoder).autocomplete({ text: 'Mill Ends' })
        expect(result).toMatchSnapshot()
      })

      it('should make search query', async () => {
        const result = await getGeocoder(geocoder).search({ text: 'Mill Ends' })
        expect(result).toMatchSnapshot()
      })

      it('should make reverse query', async () => {
        const result = await getGeocoder(geocoder)
          .reverse({ point: { lat: 45.516198, lon: -122.673240 } })
        expect(result).toMatchSnapshot()
      })
    })
  })
})
