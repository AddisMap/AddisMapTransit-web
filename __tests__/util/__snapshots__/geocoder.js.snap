// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`geocoder ARCGIS should make autocomplete query 1`] = `
Object {
  "features": Array [
    Object {
      "isCollection": false,
      "magicKey": "dHA9MCNsb2M9ODU3OTE0NyNsbmc9MzMjcGw9MzI4OTA3MSNsYnM9MTQ6MjAwNTI3NTc=",
      "text": "Mill Ends City Park, Portland, OR, USA",
    },
    Object {
      "isCollection": false,
      "magicKey": "dHA9MCNsb2M9OTY5Mzc0MiNsbmc9MzMjcGw9NTA4NTQ2NiNsYnM9MTQ6MjAwNTI3NTI=",
      "text": "Mill End Sports & Social, Penn Road, Mill End, Rickmansworth, Hertfordshire, England, WD3 8, GBR",
    },
  ],
  "query": Object {
    "text": "Mill Ends",
  },
}
`;

exports[`geocoder ARCGIS should make reverse query 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          -122.67320084756255,
          45.516175523600715,
        ],
        "type": "point",
      },
      "properties": Object {
        "country_a": "USA",
        "county": "Multnomah County",
        "label": "SW Naito Pkwy, Portland, OR, 97204, USA",
        "locality": undefined,
        "name": "SW Naito Pkwy",
        "neighbourhood": "Downtown",
        "region": "Oregon",
      },
      "type": "feature",
    },
  ],
  "query": Object {
    "lat": 45.516198,
    "lon": -122.67324,
  },
}
`;

exports[`geocoder ARCGIS should make search query 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          0.1296100000000706,
          51.973520000000065,
        ],
        "type": "point",
      },
      "properties": Object {
        "confidence": 0.875,
        "country": "GBR",
        "country_a": "GBR",
        "county": "Essex",
        "label": "Mill End, Clavering, Saffron Walden, Essex, England, GBR",
        "locality": "Saffron Walden",
        "name": "Mill End",
        "neighbourhood": "Mill End",
        "region": "England",
        "resultId": undefined,
      },
      "type": "feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -0.8668999999999301,
          51.558340000000044,
        ],
        "type": "point",
      },
      "properties": Object {
        "confidence": 0.875,
        "country": "GBR",
        "country_a": "GBR",
        "county": "Oxfordshire",
        "label": "Mill End, Hambleden, Henley-on-Thames, Oxfordshire, England, GBR",
        "locality": "Henley-on-Thames",
        "name": "Mill End",
        "neighbourhood": "Mill End",
        "region": "England",
        "resultId": undefined,
      },
      "type": "feature",
    },
  ],
  "query": Object {
    "text": "Mill Ends",
  },
}
`;

exports[`geocoder PELIAS should make autocomplete query 1`] = `
Object {
  "bbox": Array [
    -122.673377,
    45.516278,
    -122.673377,
    45.516278,
  ],
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          -122.673377,
          45.516278,
        ],
        "type": "Point",
      },
      "properties": Object {
        "accuracy": "point",
        "continent": "North America",
        "continent_gid": "whosonfirst:continent:102191575",
        "country": "United States",
        "country_a": "USA",
        "country_gid": "whosonfirst:country:85633793",
        "county": "Multnomah County",
        "county_a": "MU",
        "county_gid": "whosonfirst:county:102081631",
        "distance": 0.491,
        "gid": "openstreetmap:venue:node/4243944023",
        "id": "node/4243944023",
        "label": "Mill Ends Park, Portland, OR, USA",
        "layer": "venue",
        "locality": "Portland",
        "locality_gid": "whosonfirst:locality:101715829",
        "name": "Mill Ends Park",
        "neighbourhood": "Downtown",
        "neighbourhood_gid": "whosonfirst:neighbourhood:85867131",
        "region": "Oregon",
        "region_a": "OR",
        "region_gid": "whosonfirst:region:85688513",
        "source": "openstreetmap",
        "source_id": "node/4243944023",
      },
      "type": "Feature",
    },
  ],
  "geocoding": Object {
    "attribution": "http://lb-st-mapgeo.tri-met.org:4000/attribution",
    "engine": Object {
      "author": "Mapzen",
      "name": "Pelias",
      "version": "1.0",
    },
    "query": Object {
      "focus.point.lat": 45.52,
      "focus.point.lon": -122.67,
      "lang": Object {
        "defaulted": true,
        "iso6391": "en",
        "iso6393": "eng",
        "name": "English",
      },
      "parsed_text": Object {},
      "parser": "addressit",
      "private": false,
      "size": 10,
      "sources": Array [
        "geonames",
        "openaddresses",
        "openstreetmap",
        "whosonfirst",
      ],
      "text": "Mill Ends",
      "tokens": Array [
        "Mill",
        "Ends",
      ],
    },
    "timestamp": 1563948844191,
    "version": "0.2",
    "warnings": Array [
      "Invalid Parameter: api_key",
    ],
  },
  "isomorphicMapzenSearchQuery": Object {
    "api_key": "dummy-mapzen-key",
    "sources": "gn,oa,osm,wof",
    "text": "Mill Ends",
  },
  "type": "FeatureCollection",
}
`;

exports[`geocoder PELIAS should make reverse query 1`] = `
Array [
  Object {
    "accuracy": "point",
    "address": "Mill Ends Park, Portland, OR, USA",
    "confidence": 0.9,
    "continent": "North America",
    "continent_gid": "whosonfirst:continent:102191575",
    "country": "United States",
    "country_a": "USA",
    "country_gid": "whosonfirst:country:85633793",
    "county": "Multnomah County",
    "county_a": "MU",
    "county_gid": "whosonfirst:county:102081631",
    "distance": 0.001,
    "gid": "openstreetmap:venue:way/156366633",
    "id": "way/156366633",
    "label": "Mill Ends Park, Portland, OR, USA",
    "latlng": Object {
      "lat": 45.516201,
      "lon": -122.673245,
    },
    "layer": "venue",
    "locality": "Portland",
    "locality_gid": "whosonfirst:locality:101715829",
    "name": "Mill Ends Park",
    "neighbourhood": "Downtown",
    "neighbourhood_gid": "whosonfirst:neighbourhood:85867131",
    "region": "Oregon",
    "region_a": "OR",
    "region_gid": "whosonfirst:region:85688513",
    "source": "openstreetmap",
    "source_id": "way/156366633",
  },
  Object {
    "accuracy": "point",
    "address": "833 SW Naito Pkwy, Portland, OR, USA 97204",
    "confidence": 0.8,
    "continent": "North America",
    "continent_gid": "whosonfirst:continent:102191575",
    "country": "United States",
    "country_a": "USA",
    "country_gid": "whosonfirst:country:85633793",
    "county": "Multnomah County",
    "county_a": "MU",
    "county_gid": "whosonfirst:county:102081631",
    "distance": 0.033,
    "gid": "openaddresses:address:us/or/portland_metro:bd3f9863242e4ec7",
    "housenumber": "833",
    "id": "us/or/portland_metro:bd3f9863242e4ec7",
    "label": "833 SW Naito Pkwy, Portland, OR, USA",
    "latlng": Object {
      "lat": 45.516337,
      "lon": -122.673609,
    },
    "layer": "address",
    "locality": "Portland",
    "locality_gid": "whosonfirst:locality:101715829",
    "name": "833 SW Naito Pkwy",
    "neighbourhood": "Downtown",
    "neighbourhood_gid": "whosonfirst:neighbourhood:85867131",
    "postalcode": "97204",
    "region": "Oregon",
    "region_a": "OR",
    "region_gid": "whosonfirst:region:85688513",
    "source": "openaddresses",
    "source_id": "us/or/portland_metro:bd3f9863242e4ec7",
    "street": "SW Naito Pkwy",
  },
]
`;

exports[`geocoder PELIAS should make search query 1`] = `
Object {
  "bbox": Array [
    -122.821522,
    45.495264,
    -122.475528,
    45.516278,
  ],
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          -122.673377,
          45.516278,
        ],
        "type": "Point",
      },
      "properties": Object {
        "accuracy": "point",
        "confidence": 1,
        "continent": "North America",
        "continent_gid": "whosonfirst:continent:102191575",
        "country": "United States",
        "country_a": "USA",
        "country_gid": "whosonfirst:country:85633793",
        "county": "Multnomah County",
        "county_a": "MU",
        "county_gid": "whosonfirst:county:102081631",
        "distance": 0.491,
        "gid": "openstreetmap:venue:node/4243944023",
        "id": "node/4243944023",
        "label": "Mill Ends Park, Portland, OR, USA",
        "layer": "venue",
        "locality": "Portland",
        "locality_gid": "whosonfirst:locality:101715829",
        "match_type": "exact",
        "name": "Mill Ends Park",
        "neighbourhood": "Downtown",
        "neighbourhood_gid": "whosonfirst:neighbourhood:85867131",
        "region": "Oregon",
        "region_a": "OR",
        "region_gid": "whosonfirst:region:85688513",
        "source": "openstreetmap",
        "source_id": "node/4243944023",
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -122.681441,
          45.511851,
        ],
        "type": "Point",
      },
      "properties": Object {
        "accuracy": "centroid",
        "confidence": 1,
        "continent": "North America",
        "continent_gid": "whosonfirst:continent:102191575",
        "country": "United States",
        "country_a": "USA",
        "country_gid": "whosonfirst:country:85633793",
        "county": "Multnomah County",
        "county_a": "MU",
        "county_gid": "whosonfirst:county:102081631",
        "distance": 1.273,
        "gid": "transit:station:2584-1136::TRIMET::station",
        "id": "2584-1136::TRIMET::station",
        "label": "PSU Urban Center/SW 5th & Mill MAX Station, Portland, OR, USA",
        "layer": "station",
        "locality": "Portland",
        "locality_gid": "whosonfirst:locality:101715829",
        "match_type": "exact",
        "name": "PSU Urban Center/SW 5th & Mill MAX Station",
        "neighbourhood": "Downtown",
        "neighbourhood_gid": "whosonfirst:neighbourhood:85867131",
        "postalcode": "97201",
        "region": "Oregon",
        "region_a": "OR",
        "region_gid": "whosonfirst:region:85688513",
        "source": "transit",
        "source_id": "2584-1136::trimet::station",
        "street": "SW 5th & Mill",
      },
      "type": "Feature",
    },
  ],
  "geocoding": Object {
    "attribution": "http://lb-st-mapgeo.tri-met.org:4000/attribution",
    "engine": Object {
      "author": "Mapzen",
      "name": "Pelias",
      "version": "1.0",
    },
    "query": Object {
      "focus.point.lat": 45.52,
      "focus.point.lon": -122.67,
      "lang": Object {
        "defaulted": true,
        "iso6391": "en",
        "iso6393": "eng",
        "name": "English",
      },
      "parsed_text": Object {},
      "parser": "addressit",
      "private": false,
      "querySize": 20,
      "size": 10,
      "text": "Mill Ends",
    },
    "timestamp": 1563949031708,
    "version": "0.2",
    "warnings": Array [
      "Invalid Parameter: api_key",
    ],
  },
  "isomorphicMapzenSearchQuery": Object {
    "api_key": "dummy-mapzen-key",
    "size": 10,
    "text": "Mill Ends",
  },
  "type": "FeatureCollection",
}
`;
